### PR TITLE
82 add unit tests for quiz logic

### LINKED ISSUE #82: Add Unit Tests for Quiz logic
As a Developer, I want to test the logic of certain parts of my system in order to see how reliable it is.

- [x] JUnit is used for creating and running tests
- [x] The logic of starting a quiz is tested
    - [x] A quiz with a provided category has questions only from that category
    - [x] A theory exam has questions from every category
- [x] The logic of answering a question is tested
    - [x] Answering a question correctly increases the score by 1 // takes care of testing quiz finish logic
    - [x] Answering a question incorrectly doesn’t change the score
- [x] All tests are runnable at once
- [x] All tests pass
- [x] Tests are added to GitHub Actions

### PR DESCRIPTION
**Pull request by GeorgZs**
Closes #82 

Add all tests for logic of the quiz 
Add corner cases for tests
Update custom quiz class 
Add unit tests to CI

#### Tests are as follows:
- Test chosen category matches categories of all question
- Score increments on question 
- Review setting is done correctly and does not exceed integer limit
- Check validity of needsReview field in all questions
- Check number of questions on choosing question count option
- Check whether all possible outcomes for selecting options (category and number of questions) matches resulting quiz

### :ballot_box_with_check: Definition of Done checklist
- [x] Meaningful title and valuable description in pull request
- [x] Code change tested manually if not covered by automated tests
- [x] Pull request linked to one or more issues/stories
- [x] Acceptance criteria of the linked stories satisfied
- [x] Code change comprehended by those who approved it
- [x] State of target branch maintained or improved after the code change

### CODE REVIEW

**Comment by GeorgZs** [comment]
> that means the test would pass if for some reason the question distribution was 2 30 13. Isn't it better to check whether all categories have equal representation?

Yes that would be the case however the test aims at seeing if all 3 categories exist and not whether they each have equal spread in terms of questions @plepisnew

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/PracticeTheoryActivity.java [hunk 1]
```diff
@@ -224,7 +228,7 @@ private void addModesToModeListView() {
                 numOfQuestionsTextView.setVisibility(View.VISIBLE);
                 findViewById(R.id.cardCategory).setVisibility(View.VISIBLE);
                 MILLIS = 0;
-                numOfQuestions = 15;
+                numOfQuestions = 45;
```
Comment (Lancear) [inline SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/PracticeTheoryActivity.java:231]: maybe create a constant for 45 at the top of the file, so it is easier to find, understand, and change, than when it gets assigned on a line 231 ;)

File: SmartCarMQTTApp/app/src/test/java/com/example/smartcarmqttapp/QuizUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@RunWith(RobolectricTestRunner.class)
+public class QuizUnitTest {
+
+    //fix this either using Robolectric testing (doesnt work atm) or something else
```
Comment (Lancear) [inline SmartCarMQTTApp/app/src/test/java/com/example/smartcarmqttapp/QuizUnitTest.java]: Is this fixed or still todo?

File: SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/PracticeTheoryActivity.java (hunk 1 above)
Comment (GeorgZs) [reply SmartCarMQTTApp/app/src/main/java/com/example/smartcarmqttapp/screens/quiz/PracticeTheoryActivity.java:231]: Will do good catch!!!

File: SmartCarMQTTApp/app/src/test/java/com/example/smartcarmqttapp/QuizUnitTest.java (hunk 2 above)
Comment (GeorgZs) [reply SmartCarMQTTApp/app/src/test/java/com/example/smartcarmqttapp/QuizUnitTest.java]: Fixed, as seen in line 23: I use Robolectric

**Review by Lancear — APPROVED** [review]
image
